engines/immortal/flameSet.cpp

```diff
@@ -0,0 +1,170 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/* [Alternate Name: Torch Processing]
+ * --- What is a FlameSet ---
+ * A FlameSet is a list of torches in a given room.
+ * The level has X amount of flames total, and each
+ * room has N amount of torches, making up a FlameSet.
+ * There are 3 instances of torches in the game.
+ * First is in the story record, second is in the level torches,
+ * and lastly is the in-room torch. This is done so that
+ * the in-room torch can be lit up by a fireball and then
+ * stay permanantly lit for the next time the player enters the room.
+ */
+
+#include "immortal/room.h"
+
+namespace Immortal {
+
+void Room::flameInit() {
+	flameFreeAll();
+	_candleTmp = 0;
+}
+
+void Room::flameFreeAll() {
+	_numFlames = 0;
+	_numInRoom = 0;
+}
+
+void Room::flameDrawAll(uint16 vX, uint16 vY) {
+	for (int i = 0; i < _fset.size(); i++) {
+		// For every flame in the room, add the sprite to the sprite table
+		univAddSprite(vX, vY, _fset[i]._x, _fset[i]._y, g_immortal->_cycPtrs[g_immortal->_cycles[_fset[i]._c]._cycList]._sName, cycleGetFrame(_fset[i]._c), 0);
+		if (cycleAdvance(_fset[i]._c) == true) {
+			cycleFree(_fset[i]._c);
+			_fset[i]._c = flameGetCyc(&_fset[i], 1);
+		}
+	}
+}
+
+bool Room::roomLighted() {
+	// For testing purposes, we say it's always lit
+	return true;
+
+	// Very simple, just checks every torch and if any of them are lit, we say the room is lit
+	for (int i = 0; i < _fset.size(); i++) {
+		if (_fset[i]._p != kFlameOff) {
+			return true;
+		}
+	}
+	return false;
+}
+
+void Room::lightTorch(uint8 x, uint8 y) {
+	/* Checks every torch to see if it is:
+	 * pattern == off, and inside the point x,y
+	 * which is the fireball position. This is a
+	 * little bit clever, because it saves cycles
+	 * over checking x,y first, since you need to
+	 * check both x,y and flame pattern. Neato.
+	 */
+
+	for (int i = 0; i < _fset.size(); i++) {
+		if (_fset[i]._p == kFlameOff) {
+			if (Utilities::inside(kLightTorchX, x, y, _fset[i]._x + 16, _fset[i]._y + 8)) {
+				_fset[i]._p = kFlameNormal;
+
+			}
+		}
+	}
+}
+
+void Room::flameSetRoom(Common::Array<SFlame> allFlames) {
```

Comment:
Probably it is better to pass by reference, e.g. `Common::Array<SFlame> &allFlames`, so the list is not copied every time.